crates/wash-cli/src/common/registry_cmd.rs

```diff
@@ -30,19 +35,54 @@ pub async fn registry_pull(
     cmd: RegistryPullCommand,
     output_kind: OutputKind,
 ) -> Result<CommandOutput> {
-    let artifact_url = cmd.url.to_ascii_lowercase();
-    let image: Reference = artifact_url.parse()?;
+    let artifact_url = match cmd.url {
+        Some(url) => Ok(url),
+        None => {
+            let project_config = get_config(None, Some(true)).map_err(|e| {
+                anyhow!(
+                    "Failed to read project config to extract a registry URL: {}",
+                    e
+                )
+            })?;
+
+            project_config.resolve_registry_url().map_err(|_| {
+                anyhow!(
+                    "No registry URL configured. Please specify a registry URL with the --url flag"
+                )
+            })
+        }
+    }?;
+
+    let image: Reference = artifact_url.to_ascii_lowercase().parse()?;
 
     let spinner = Spinner::new(&output_kind)?;
     spinner.update_spinner_message(format!(" Downloading {} ...", image.whole()));
 
+    let credentials = match (cmd.opts.user, cmd.opts.password) {
+        (Some(user), Some(password)) => Ok(RegistryAuth::Basic(user, password)),
+        _ => match get_config(None, Some(true)) {
+            Ok(project_config) => {
+                project_config
+                    .resolve_registry_credentials(&artifact_url)
+                    .await
+            }
+            Err(_) => Ok(RegistryAuth::Anonymous),
+        },
+    }?;
+
     let artifact = pull_oci_artifact(
         artifact_url,
         OciPullOptions {
             digest: cmd.digest,
             allow_latest: cmd.allow_latest,
-            user: cmd.opts.user,
-            password: cmd.opts.password,
+            user: match &credentials {
```

Comment:
Nit: since we're checking the same condition twice, you could assign to `(user, password)` above this block